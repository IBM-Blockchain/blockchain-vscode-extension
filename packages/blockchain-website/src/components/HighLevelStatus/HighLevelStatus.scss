.current-status {
    font-size: 14px;
    color: #161616;
    letter-spacing: 0.16px;
    line-height: 18px;
}

.high-level-status {
    border-radius: 4px;
    display: inline;
    background-color: #b5b5b5;
    color: #fff;
    font-size: 12px;
    font-size: 0.75rem;
    line-height: 2;
    font-weight: bold;
    padding: 3px 7px;
    text-align: center;
    vertical-align: baseline;
    white-space: nowrap;
}

.high-level-status:hover {
    cursor: pointer;
}

.bg-red {
    background-color: red !important;
}
  
.bg-amber {
    background-color: orange !important;
}

.bg-green {
    background-color: green !important;
}

.icon {
    position: fixed;
    cursor: pointer;
    width: 20px;
    height: 20px;
    padding-left: 3px;
}