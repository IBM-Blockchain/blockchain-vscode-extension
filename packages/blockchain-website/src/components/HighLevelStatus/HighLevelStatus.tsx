import React, { Component } from 'react';
import Axios, { AxiosResponse } from 'axios';
import './HighLevelStatus.scss';
import checkmark from '../../checkmark.svg';
import error from '../../error.svg';
import warning from '../../warning.svg';

const githubProject: string = 'IBM-Blockchain/blockchain-vscode-extension';

// Github URLs
const issuesStatusApiUrl: string = 'https://api.github.com/repos/' + githubProject + '/issues?labels=status';

interface HighLevelStatusState {
    description: string;
    icon: string;
}

class HighLevelStatus extends Component<{}, HighLevelStatusState> {

    constructor(props: Readonly<HighLevelStatusState>) {
        super(props);
        this.state = {
            description: '',
            icon: ''
        };
    }

    async componentDidMount(): Promise<void> {
        let description: string;
        let icon: string;
        let sev1Count: number = 0;
        let sev2Count: number = 0;
        let sev3Count: number = 0;

        try {
            const response: AxiosResponse = await Axios.get(issuesStatusApiUrl);
            const allIssues: any = response.data;
            allIssues.forEach((issue: any) => {
                if (issue.labels) {
                    issue.labels.forEach((label: {id: number, node_id: string, url: string, name: string, color: string, default: boolean, description: string}) => {
                        if (label.name === 'sev1') {
                            sev1Count += 1;
                        } else if (label.name === 'sev2') {
                            sev2Count += 1;
                        } else if (label.name === 'sev3') {
                            sev3Count += 1;
                        }
                    });
                }
            });
            if (sev1Count > 0) {
                icon = error;
                description = 'There is at least one sev1 issue';
            } else if (sev2Count > 0 || sev3Count > 0) {
                icon = warning;
                description = 'There is at least one sev2 or sev3 issue';
            } else {
                icon = checkmark;
                description = 'There are no known issues';
            }

            this.setState({
                description: description,
                icon: icon
            });
        } catch (error) {
            return;
        }
    }

    render(): JSX.Element {
        if (this.state.icon !== '') {
            return(
                <p className='current-status'>
                    <span>
                        Current overall status: <span> </span> {/* Added a space after colon using span */}
                    </span>
                    <img className='icon' alt='' title={this.state.description} src={this.state.icon}/>
                </p>
            );
        } else {
            return <div/>;
        }
    }
}

export default HighLevelStatus;
