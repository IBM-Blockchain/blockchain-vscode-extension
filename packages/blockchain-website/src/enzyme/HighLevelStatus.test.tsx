import React from 'react';
import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import chai from 'chai';
import sinon from 'sinon';
import Axios from 'axios';
import sinonChai from 'sinon-chai';
import HighLevelStatus from '../components/HighLevelStatus/HighLevelStatus';
chai.should();
chai.use(sinonChai);

const githubProject: string = 'IBM-Blockchain/blockchain-vscode-extension';

// Github URLs
const issuesStatusApiUrl: string = 'https://api.github.com/repos/' + githubProject + '/issues?labels=status';

describe('HighLevelStatus', () => {
    let mySandBox: sinon.SinonSandbox;
    let axiosGetStub: sinon.SinonStub;

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();
        axiosGetStub = mySandBox.stub(Axios, 'get');
        axiosGetStub.returns({
            data: [
                {
                    url: 'someURL',
                    title: 'someTitle',
                    labels: [
                        {
                            id: 'someLabelID',
                            name: 'sev1'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                },
                {
                    url: 'someOtherURL',
                    title: 'someOtherTitle',
                    labels: [
                        {
                            id: 'someLabelID',
                            name: 'sev2'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                },
                {
                    url: 'someOtherURL1',
                    title: 'someOtherTitle1',
                    labels: [
                        {
                            id: 'someLabelID1',
                            name: 'sev3'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                },
                {
                    url: 'anotherURL',
                    title: 'anotherTitle',
                    labels: [
                        {
                            id: 'anotherLabelID',
                            name: 'sev1'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                }
            ]
        });
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    it('should render the expected snapshot', async () => {
        const component: any = renderer
            .create(<HighLevelStatus />)
            .toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should output current status when successful api call and no incidents', async () => {
        axiosGetStub.returns({
            data: [
                {}
            ]});
        const component: ReactWrapper<{}, {description: string, icon: string}, HighLevelStatus> = await mount(<HighLevelStatus/>);
        axiosGetStub.should.have.been.calledOnceWithExactly(issuesStatusApiUrl);
        component.state().description.should.equal('There are no known issues');
    });

    it('should output current status when successful api call and sev2 or sev3 incidents', async () => {
        axiosGetStub.returns({
            data: [
                {
                    url: 'someOtherURL',
                    title: 'someOtherTitle',
                    labels: [
                        {
                            id: 'someLabelID',
                            name: 'sev2'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                },
                {
                    url: 'someOtherURL1',
                    title: 'someOtherTitle1',
                    labels: [
                        {
                            id: 'someLabelID1',
                            name: 'sev3'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                }
            ]});
        const component: ReactWrapper<{}, {description: string, icon: string}, HighLevelStatus> = await mount(<HighLevelStatus/>);
        axiosGetStub.should.have.been.calledOnceWithExactly(issuesStatusApiUrl);
        component.state().description.should.equal('There is at least one sev2 or sev3 issue');
    });

    it('should output current status when successful api call and sev1 incident', async () => {
        axiosGetStub.returns({
            data: [
                {
                    url: 'someURL',
                    title: 'someTitle',
                    labels: [
                        {
                            id: 'someLabelID',
                            name: 'sev1'
                        },
                        {
                            id: 'someOtherLabelID',
                            name: 'status'
                        }
                    ]
                }
            ]});
        const component: ReactWrapper<{}, {description: string, icon: string}, HighLevelStatus> = await mount(<HighLevelStatus/>);
        axiosGetStub.should.have.been.calledOnceWithExactly(issuesStatusApiUrl);
        component.state().description.should.equal('There is at least one sev1 issue');
    });

    it('should output worst current status when successful api call and there multiple different severity incidents', async () => {
        const component: ReactWrapper<{}, {description: string, icon: string}, HighLevelStatus> = await mount(<HighLevelStatus/>);
        axiosGetStub.should.have.been.calledOnceWithExactly(issuesStatusApiUrl);
        component.state().description.should.equal('There is at least one sev1 issue');
    });

    it('should output nothing when unsuccessful api call', async () => {
        axiosGetStub.returns(undefined);
        const component: ReactWrapper<{}, {description: string, icon: string}, HighLevelStatus> = await mount(<HighLevelStatus/>);
        axiosGetStub.should.have.been.calledOnceWithExactly(issuesStatusApiUrl);
        component.state().description.should.equal(''); // i.e. an icon/status was not shown to the user
    });
});
