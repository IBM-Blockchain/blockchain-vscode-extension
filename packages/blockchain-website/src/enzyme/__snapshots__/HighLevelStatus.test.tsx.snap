// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighLevelStatus should render the expected snapshot 1`] = `<div />`;
